### ISSUE TITLE
lock user account to serial console only

### ISSUE THREAD

**Issue by breyn407** [Microsoft/azserialconsole#60]
In order to use serial console, I apparently have to cripple my entire app's security posture by creating a login account with a password instead of an ssh key only.

The serial console is essentially a last-resort tool if I lose ssh.  Since I never plan on losing ssh (who does?), the only way this is an effective tool is if I leave a password only login account enabled for ssh access active 24/7 (i.e. ALL the time).  This would allow access to the serial console in those unplanned situations.

Do you see the problem here?  I have to choose between security and this tool.  For production, this is a no-brainer.  This tool is absolutely off-limits for production (i.e. useless).  Even for dev, I don't like leaving my backside hanging out as a requirement for using this tool.  But dev is where this is really needed because that's where you have the freedom to muck about.  That is where you need a last-resort recovery tool.

If the app is not public-facing, then I suppose it is not as big of a deal.  So, don't use Azure IaaS linux for public-facing solutions.  Got it.

The solution seems simple (besides the obvious one of providing a mechanism to use an RSA key for login):  How can I restrict a linux user account to only provide a shell for the serial console?

**Comment by asinn826** [comment]
Hey @breyn407 - thanks for the feedback! We have key-based logins as something we want to explore further, the challenge for us is that the serial connection itself does not allow SSH so we can't do key-pair authentication at that level. I suppose one alternative would be to do that key-pair check at the Azure portal level before you can even load the serial console blade - what are your thoughts?

Your suggestion on restricting a Linux user account to provide only a shell for serial console is interesting. Were you thinking that you might have a separate password auth-based user account that is only allowed to communicate on ttyS0? Would love to hear your thoughts on that some more too.

**Comment by breyn407** [comment]
@asinn826, yes!  That's exactly what I was thinking.  To restrict a password-only user account (newly created for just this purpose) to ttyS0.  I was thinking of something along the lines of:

edit /etc/security/access.conf

+:foobar:ttyS0
-:foobar:ALL

User foobar is the user restricted to only ttyS0 login.  Let me know your thoughts.  I'll have to test.

**Comment by asinn826** [comment]
@breyn407 I got a chance to try out what you suggested on a RHEL 7.6 box, and it did work. I just had to do a couple more modifications in addition to the /etc/security/access.conf change that you mentioned:

1. Modified my /etc/pam.d/login to look like the following (added the `pam_access.so` line in the first section):
```#%PAM-1.0                                                                                   
auth [user_unknown=ignore success=ok ignore=ignore default=bad] pam_securetty.so            
auth       substack     system-auth                                                         
auth       include      postlogin                                                           
account    required     pam_nologin.so                                                      
account    required     pam_access.so                                                       
account    include      system-auth                                                         
password   include      system-auth                                                         
# pam_selinux.so close should be the first session rule                                     
session    required     pam_selinux.so close                                                
session    required     pam_loginuid.so                                                     
session    optional     pam_console.so                                                      
# pam_selinux.so open should only be followed by sessions to be executed in the user context
session    required     pam_selinux.so open                                                 
session    required     pam_namespace.so                                                    
session    optional     pam_keyinit.so force revoke                                         
session    include      system-auth                                                         
session    include      postlogin                                                           
-session   optional     pam_ck_connector.so```

This allows pam to scan /etc/security/access.conf for rules.

2. Modify my /etc/ssh/sshd_config to have a rule to disallow ssh for my test user:
```DenyUsers testuser```

I *think* that should work (at least it seemed to work on my test VM)...but if you get a chance, try it out and let me know what you see!

**Comment by Klaas-** [comment]
I think the easiest way to achieve this is to use root for last-resort recovery purposes with a password and then set ssh to not allow root to login / without password only. `PermitRootLogin prohibit-password` which is the default -- or `PermitRootLogin no`; this does assume that you do not have additional services where you can use local users to login -- like for example cockpit
